build/sdk_update.go

```diff
@@ -0,0 +1,316 @@
+		return c.processTemplateAsset(name, c.context, overwrite)
+	}
+
+	return c.processFileAsset(name, overwrite)
+}
+
+func (c *updater) processFileAsset(name string, overwrite bool) error {
+	content := skeleton.MustAsset(name)
+	info, _ := skeleton.AssetInfo(name)
+
+	name = fixGitFolder(name)
+	return c.writeFile(filepath.Join(c.root, name), content, info.Mode(), overwrite)
+}
+
+var funcs = map[string]interface{}{
+	"escape_shield": escapeShield,
+	"expName":       expName,
+}
+
+func (c *updater) processTemplateAsset(name string, v interface{}, overwrite bool) error {
+	tpl := string(skeleton.MustAsset(name))
+
+	t, err := template.New(name).Funcs(funcs).Parse(tpl)
+	if err != nil {
+		return err
+	}
+
+	name = fixGitFolder(name)
+	file := filepath.Join(c.root, name[:len(name)-len(tplExt)])
+
+	buf := bytes.NewBuffer(nil)
+	if err := t.Execute(buf, v); err != nil {
+		return err
+	}
+
+	info, _ := skeleton.AssetInfo(name)
+	return c.writeFile(file, buf.Bytes(), info.Mode(), overwrite)
+}
+
+func (c *updater) writeFile(file string, content []byte, m os.FileMode, overwrite bool) error {
+	f, err := os.Open(file)
+	if err != nil && !os.IsNotExist(err) {
+		return err
+	}
+
+	if f == nil {
+		c.notifyMissingFile(file)
+		return c.doWriteFile(file, content, m)
+	}
+
+	if !overwrite {
+		return nil
+	}
+
+	original, err := ioutil.ReadAll(f)
+	if err != nil {
+		return err
+	}
+
+	if bytes.Compare(original, content) == 0 {
+		return nil
+	}
+
+	c.notifyChangedFile(file)
+	return c.doWriteFile(file, content, m)
+}
+
+func (c *updater) doWriteFile(file string, content []byte, m os.FileMode) error {
+	if c.opt.DryRun {
+		return nil
+	}
+
+	if err := os.MkdirAll(filepath.Dir(file), 0755); err != nil {
+		return err
+	}
+
+	f, err := os.OpenFile(file, os.O_RDWR|os.O_CREATE|os.O_TRUNC, m)
+	if err != nil {
+		return err
+	}
+
+	defer f.Close()
+
+	if c.opt.Debugf != nil {
+		c.opt.Debugf("file %q has been written\n", file)
+	}
+
+	_, err = f.Write(content)
+	if err != nil {
+		return err
+	}
+	rel, err := filepath.Rel(c.root, file)
+	if err != nil {
+		return err
+	}
+	if !strings.HasPrefix(rel, ".git"+string(filepath.Separator)) {
+		git := exec.Command("git", "add", rel)
+		git.Dir = c.root
+		if out, err := git.CombinedOutput(); err != nil {
+			c.opt.Warningf("cannot add a file to git: %v\n%s", err, string(out))
+		}
+	}
+	return nil
+}
+
+func (c *updater) notifyMissingFile(file string) {
+	if !c.opt.DryRun {
+		c.opt.Noticef("creating file %q\n", file)
+		return
+	}
+
+	if isDotGit(file) {
+		return
+	}
+
+	c.changes++
+	c.opt.Warningf("missing file %q\n", file)
+}
+
+func (c *updater) notifyChangedFile(file string) {
+	if !c.opt.DryRun {
+		c.opt.Warningf("managed file %q has changed, overriding changes\n", file)
+		return
+
+	}
+
+	c.changes++
+	c.opt.Warningf("managed file changed %q\n", file)
+}
+
+func escapeShield(text interface{}) string {
+	return strings.Replace(fmt.Sprintf("%s", text), "-", "--", -1)
+}
+
+func expName(s string) string {
```

Comment:
What does the name of this function mean? It looks like it capitalizes the first rune in the string, but the name doesn't suggest that.